### PR TITLE
Added some needed things to attention, implemented first encoder, dec…

### PR DESCRIPTION
**Pull request by mpintaric55334**
This PR is not finished, but it may have some war crimes in torch malpractices, so better to check it now before it is massacred further. I think I did some stupid things with layer norm and dropout, that I use it too much, will ask that on the next meet. Also, the masking for axial attention creates unique problem(s), will present it on meeting, I have some fixes ready.

### CODE REVIEW

**Comment by mpintaric55334** [comment]
Attention nan problem identified and fixed

**Comment by mpintaric55334** [comment]
Added ignoring of padding in encoder output

File: model/attention.py [hunk 1]
```diff
@@ -12,9 +12,16 @@ def dot_product_attention(q, k, v, attn_mask=None):
     attn = torch.matmul(q, k.transpose(-1, -2)) / math.sqrt(c)
 
     if attn_mask is not None:
-        attn = attn.masked_fill(attn_mask, float("-inf"))
-
+        attn = attn.masked_fill(attn_mask == 0, float("-1e34"))
```
Comment (RJPenic) [inline model/attention.py]: I think you don't need `float(...)` when defining `-1e34`.

File: model/encoder.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+class EncoderOutputReduction:
+
+    """
+    Class that implements the final reduction of encoder.
+    After passing through the encoder block, its output
+    is of shape (batch_size, N, N, embedd_size).
+    This is not compatible
+    """
+
+    def __init__(self, reduction: str):
```
Comment (RJPenic) [inline model/encoder.py]: Is this correct? Argument reduction is `str`?

File: model/decoder.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                                            dropout=dropout,
+                                            attention_dropout=attention_dropout,
+                                            bias=attention_bias)
+        self.cross_att = AttentionLayer(embedd_size=embedd_size,
+                                        num_heads=num_heads,
+                                        dropout=attention_dropout,
+                                        attention_dropout=attention_dropout,
+                                        bias=attention_bias)
+        self.fc_layer = FeedForwardLayer(fc_dim=embedd_size, bias=fc_bias,
+                                         dropout=fc_dropout)
```
Comment (RJPenic) [inline model/decoder.py]: Decoder consists of only one block? They usually have multiple block (similar to encoder). Is this how they did it in RNAinformer?

File: model/decoder.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    def forward(self, dec_input, enc_output, attn_mask=None):
```
Comment (RJPenic) [inline model/decoder.py]: We'll probably need two types of masks for the decoder. Check `key_padding_mask `and `attn_mask` [here](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html). In decoder you need to mask out "future" tokens.

File: model/encoder.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        where embedding dim is actually channels??,
+        whats the deal with that, is there some
+        paper or article which explains this in more
+        detail, or is this one of it works, we dont know why things
+        in ML
+        """
+        _x = x
+        x = x.permute(0, 3, 1, 2)
+        # does normalisation go before or after i remeber residual connections?
+        x = self.norm1(x)
```
Comment (RJPenic) [inline model/encoder.py:66]: Some people use it before, some after. Doubt it will affect results too much. Leave it like it is for now, we can easily try changing it later.

File: model/encoder.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        eps = 1e-34
+
+        if mask is not None:
+            # before unsqueeze mask shape (bs,N,N)
+            mask = mask.unsqueeze(-1)
+            # after shape is (bs,N,N,1), compatible for pointwise multip
+            x = x * mask
+
+        x = torch.sum(x, self.reduction) / (eps + torch.sum(mask,
+                                                            self.reduction))
```
Comment (RJPenic) [inline model/encoder.py]: This will throw an error if mask is `None`.

File: model/encoder.py (hunk 2 above)
Comment (RJPenic) [reply model/encoder.py]: If I understood correctly, it is either an `int` or `tuple`? In that case you could use `reduction: int | tuple[int, int]`.

File: model/encoder.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        super(EncoderBlock, self).__init__()
+        self.attention_layer = AxialAttentionLayer(embedd_size, num_heads,
+                                                   attention_dropout,
+                                                   bias=bias)
+        self.convolution_layer = ConvLayer(embed_size=embedd_size,
+                                           kernel_size=kernel_size_conv,
+                                           out_channels=conv_out_channels,
+                                           padding=padding)
+        # should I add another dropout?
```
Comment (RJPenic) [inline model/encoder.py:160]: Should be okay without it, we can add it if we see that model overfits.

File: model/encoder.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        In RNAInformer they divide dropout by 2 for
+        row and column attention, is that the mainstream way
+        of distributing dropout by rows and columns in axial??
+        Also, if I have dropout in my attention class is it even necessary here
+        (i have same dropout inside attention class and after class output)
+        """
+        super(AxialAttentionLayer, self).__init__()
+        self.dropout1 = nn.Dropout(dropout / 2)
+        self.dropout2 = nn.Dropout(dropout / 2)
```
Comment (RJPenic) [inline model/encoder.py:99]: Why is this divided by two?

File: model/encoder.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+             only same works
+        """
+        super(EncoderBlock, self).__init__()
+        self.attention_layer = AxialAttentionLayer(embedd_size, num_heads,
+                                                   attention_dropout,
+                                                   bias=bias)
+        self.convolution_layer = ConvLayer(embed_size=embedd_size,
+                                           kernel_size=kernel_size_conv,
+                                           out_channels=conv_out_channels,
+                                           padding=padding)
```
Comment (RJPenic) [inline model/encoder.py:159]: I assume this doesn't reduce length dimensions in any way?

**Review by RJPenic — DISMISSED** [review]

File: model/decoder.py (hunk 3 above)
Comment (mpintaric55334) [reply model/decoder.py]: My bad, this was supposed to be DecoderBlock, will add it to decoder

File: model/encoder.py (hunk 8 above)
Comment (mpintaric55334) [reply model/encoder.py:99]: They did it like that in RNAinformer, one dropout is for row attention, one for column, so I guess they split the initial dropout factor by 2 to distribute it evenly

File: model/encoder.py (hunk 9 above)
Comment (mpintaric55334) [reply model/encoder.py:159]: Nope, remains the same

**Comment by mpintaric55334** [comment]
Model completely done, I hope so.
